src/utils/src/ThreadsafeBlockingQueue.c

```diff
@@ -0,0 +1,214 @@
+#include "Include_i.h"
+
+/**
+ * Create a threadsafe blocking queue
+ */
+STATUS safeBlockingQueueCreate(PSafeBlockingQueue* ppSafeQueue)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    PSafeBlockingQueue pSafeQueue = NULL;
+
+    CHK(ppSafeQueue != NULL, STATUS_NULL_ARG);
+
+    // Allocate the main structure
+    pSafeQueue = (PSafeBlockingQueue) MEMCALLOC(1, SIZEOF(SafeBlockingQueue));
+    CHK(pSafeQueue != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    ATOMIC_STORE_BOOL(&pSafeQueue->terminate, FALSE);
+
+    pSafeQueue->mutex = MUTEX_CREATE(FALSE);
+    CHK_STATUS(semaphoreEmptyCreate(INT32_MAX, &(pSafeQueue->semaphore)));
+    CHK_STATUS(stackQueueCreate(&(pSafeQueue->queue)));
+
+    *ppSafeQueue = pSafeQueue;
+
+CleanUp:
+    if (STATUS_FAILED(retStatus) && pSafeQueue != NULL) {
+        SAFE_MEMFREE(pSafeQueue);
+    }
+
+    return retStatus;
+}
+
+/**
+ * Frees and de-allocates the thread safe blocking queue
+ */
+STATUS safeBlockingQueueFree(PSafeBlockingQueue pSafeQueue)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+
+    // set terminate flag, lock mutex -- this assures all other threads
+    // are no longer directly interacting with the queue
+    //
+    // free semaphore, this will unblock any threads blocking on said semaphore,
+    // and then exit from the terminate flag
+    //
+    // unlock mutex
+    ATOMIC_STORE_BOOL(&pSafeQueue->terminate, TRUE);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    CHK_STATUS(semaphoreFree(&(pSafeQueue->semaphore)));
+    MUTEX_UNLOCK(pSafeQueue->mutex);
+
+    CHK_STATUS(stackQueueFree(pSafeQueue->queue));
+    MUTEX_FREE(pSafeQueue->mutex);
+
+    SAFE_MEMFREE(pSafeQueue);
+
+CleanUp:
+
+    return retStatus;
+}
+
+/**
+ * Clears and de-allocates all the items
+ */
+STATUS safeBlockingQueueClear(PSafeBlockingQueue pSafeQueue, BOOL freeData)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    // 0 timeout semaphore acquire, eventually getting a timeout, this is done to clear
+    // the counting semaphore
+    while ((semaphoreAcquire(pSafeQueue->semaphore, 0) == STATUS_SUCCESS) && !ATOMIC_LOAD_BOOL(&pSafeQueue->terminate))
```

Comment:
This will peg the CPU core to 100%. There needs to be a short delay to create spinlock type of thing - something like 5-30milliseconds.

Reply:
Not quite. In the event the semaphore count is reduced to 0 this operation won't return to STATUS_SUCCESS, it'll return some variation of STATUS_TIMEOUT, which is what we want. A short delay just slows down this finite operation.

Reply:
Oh I see, I misread it to be not equal to STATUS_SUCCESS

Reply:
With setting permit count to 0, this will actually be an infinite loop unless pSafeQueue->terminate is set to TRUE while freeing the queue. Check comment above.

Reply:
Not quite. When the `existingCount <= 0` the code proceeds to line ~198, this line:

`CHK_STATUS(CVAR_WAIT(pSemaphore->semaphoreNotify, pSemaphore->semaphoreLock, timeout));`
Here the timeout is 0, so it will immediately timeout and the status won't be STATUS_SUCCESS, it will be some variation of STATUS_TIMEOUT.